stl/src/filesystem.cpp

```diff
@@ -665,6 +675,51 @@ _Success_(return == __std_win_error::_Success) __std_win_error
 
 [[nodiscard]] __std_win_error __stdcall __std_fs_rename(
     _In_z_ const wchar_t* const _Source, _In_z_ const wchar_t* const _Target) noexcept {
+    __std_win_error _Last_error;
+    constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;

```

Comment:
Does `__std_fs_file_flags::_Backup_semantics` work under a limited user account?

Reply:
Not sure, but we need this to rename a directory.

Reply:
I don't think `MoveFileEx` uses this flag internally, as it can use `NtOpenFile` directly.

My concern is about this part of [CreateFileW](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew) doc:

> FILE_FLAG_BACKUP_SEMANTICS
> 0x02000000
> The file is being opened or created for a backup or restore operation. The system ensures that the calling process overrides file security checks when the process has SE_BACKUP_NAME and SE_RESTORE_NAME privileges. For more information, see Changing Privileges in a Token.

If there's a chance that there's dependency on privilege, the function may behave inconsistently, depending on that.

Reply:
Do we have a way to get a `HANDLE` to a directory without `FILE_FLAG_BACKUP_SEMANTICS`? I am not sure about using the Nt* stuff. Is it allowed here?

Reply:
I don't know. Undocumented stuff is not allowed (#705). But maybe NtOpenFile is fine, there's a documentation https://docs.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntopenfile

Would be good if just using `FILE_FLAG_BACKUP_SEMANTICS` is fine, but need to confirm it.

Reply:
`FILE_FLAG_BACKUP_SEMANTICS` can be used by limited users; we already use it elsewhere in this implementation. The documentation is somewhat confusing, what it means is that if you have taken appropriate privileges, and you use the flag, then you get the effects of those privileges (ignoring read restrictions with SeBackupPrivilege, and write restrictions with SeRestorePrivilege). If you haven't taken those privileges then you still get the ability to open a directory and other effects of the flag, you just don't get to ignore ACLs.

Reply:
> I don't know. Undocumented stuff is not allowed (#705). But maybe NtOpenFile is fine, there's a documentation https://docs.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntopenfile

`NtOpenFile` is not available under the "app"/"onecore" flavors, to my understanding.

Reply:
What would be nice is to be able to use `NtSetInformationFile`. It's documented in the ddk https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntsetinformationfile. Is this allowed?

With this we can also fix a similar issue(due to read-only bit) in `filesystem::create_hard_link`